### PR TITLE
[AuthV2] Nov 2022 Bug Fixes and Improvements

### PR DESCRIPTION
**Pull request by mkarmark**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Review by cgillum — APPROVED** [review]
Looks mostly good, though we may want to see if we can add a few more tests to validate the various parameter combinations you're introducing. Otherwise, really happy to see this PR!!

**Comment by mkarmark** [comment]
@zhoxing-ms do you have any idea why my integration tests are failing? They seem to be failing at the webapp create step which doesn't have anything to do with my code.

**Comment by cgillum** [comment]
Adding James (@Hamster-Huey) for FYI.

**Comment by yonzhan** [comment]
AuthV2

File: src/authV2/azext_authV2/custom.py [hunk 1]
```diff
@@ -452,11 +462,39 @@ def get_aad_settings(cmd, resource_group_name, name, slot=None):
...
+                       'configured to non empty strings')
+
+    if client_secret_certificate_thumbprint is not None and client_secret_certificate_san is not None:
+        raise CLIError('Usage Error: --thumbprint and --san cannot both be '
+                       'configured to non empty strings')
+
+    if ((client_secret_certificate_san is not None and client_secret_certificate_issuer is None) or
+            (client_secret_certificate_san is None and client_secret_certificate_issuer is not None)):
+        raise CLIError('Usage Error: --san and --certificate-issuer must both be '
+                       'configured to non empty strings')
```
Comment (zhoxing-ms) [inline src/authV2/azext_authV2/custom.py:496]: Could you please use the specific error type from `azure.cli.core.azclierror` instead of `CLIError`?

**Comment by zhoxing-ms** [comment]
Could you add some new tests for those changes?

**Comment by zhoxing-ms** [comment]
@mkarmark Please resolve those CI issues~

**Comment by zhoxing-ms** [comment]
Please write the description of those changes into `HISTORY.rst`. In addition, if you want to upgrade the new version to release these features, please modify `setup.py` at the same time

**Comment by mkarmark** [comment]
> @mkarmark Please resolve those CI issues~

Do you have any idea why those two CI tests could be failing? They seem to be failing upstream of my extension code so I'm a little bit confused about how to fix it. Also everything passes when I run them locally on my machine and I have also tried re-recording them.

**Comment by zhoxing-ms** [comment]
@mkarmark Please note that there are problems with the following two tests:
- test_authV2_auth (azext_authV2.tests.latest.test_authV2_scenario.Authv2ScenarioTest)
- test_authV2_authclassic (azext_authV2.tests.latest.test_authV2_scenario.Authv2ScenarioTest)

Error details:
> File "/home/vsts/work/1/s/azure-cli/src/azure-cli/azure/cli/command_modules/appservice/custom.py", line 1165, in _fill_ftp_publishing_url
    profiles = list_publish_profiles(cmd, resource_group_name, name, slot)
  File "/home/vsts/work/1/s/azure-cli/src/azure-cli/azure/cli/command_modules/appservice/custom.py", line 2076, in list_publish_profiles
    for f in content:
  File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/azure/core/pipeline/transport/_requests_basic.py", line 182, in __next__
    raise IncompleteReadError(err, error=err)
azure.core.exceptions.IncompleteReadError: ('Connection broken: IncompleteRead(1771 bytes read, 104 more expected)', IncompleteRead(1771 bytes read, 104 more expected))

**Comment by mkarmark** [comment]
@zhoxing-ms  I did rerecord the tests with live mode, and as you can see in the errors the error is coming from the webapp create step which I didn't touch in this entire extension.

**Comment by zhoxing-ms** [comment]
> File "/home/vsts/work/1/s/azure-cli/src/azure-cli/azure/cli/command_modules/appservice/custom.py", line 1166, in _fill_ftp_publishing_url
    profiles = list_publish_profiles(cmd, resource_group_name, name, slot)
  File "/home/vsts/work/1/s/azure-cli/src/azure-cli/azure/cli/command_modules/appservice/custom.py", line 2085, in list_publish_profiles
    for f in content:
  File "/opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/azure/core/pipeline/transport/_requests_basic.py", line 182, in __next__
    raise IncompleteReadError(err, error=err)
azure.core.exceptions.IncompleteReadError: ('Connection broken: IncompleteRead(1771 bytes read, 104 more expected)', IncompleteRead(1771 bytes read, 104 more expected))


There is an `IncompleteReadError` with `full_xml += f.decode()` in method `list_publish_profiles()`, which blocks the CI of this PR. @panchagnula Could you please help to have a look~
[code link](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/appservice/custom.py#L2090)

**Comment by zhoxing-ms** [comment]
Know issue, it is the same issue as https://github.com/Azure/azure-cli/pull/20459#discussion_r758076069

**Comment by mkarmark** [comment]
@zhoxing-ms What's the ETA for the fix for this issue?

**Comment by jsntcy** [comment]
@mkarmark, could you please merge `Main` branch into your branch to see if CI passes?

**Comment by mkarmark** [comment]
@jsntcy The tests are still failing.

**Comment by Hamster-Huey** [comment]
@jiasli @jsntcy looks like all checks pass now, can you review again

**Comment by yonzhan** [comment]
Add @zhoxing-ms to help with

File: src/authV2/azext_authV2/custom.py [hunk 2]
```diff
@@ -452,14 +463,42 @@ def get_aad_settings(cmd, resource_group_name, name, slot=None):
...
+    if ((client_secret_certificate_san is not None and client_secret_certificate_issuer is None) or
+            (client_secret_certificate_san is None and client_secret_certificate_issuer is not None)):
+        raise AzCLIError('Usage Error: --san and --certificate-issuer must both be '
+                         'configured to non empty strings')
 
     if issuer is not None and (tenant_id is not None):
-        raise CLIError('Usage Error: --issuer and --tenant-id cannot be configured '
-                       'to non empty strings at the same time.')
+        raise AzCLIError('Usage Error: --issuer and --tenant-id cannot be configured '
+                         'to non empty strings at the same time.')
```
Comment (zhoxing-ms) [inline src/authV2/azext_authV2/custom.py:501]: Could you please use `ArgumentUsageError` instead of `AzCLIError`?

Comment (Hamster-Huey) [reply src/authV2/azext_authV2/custom.py]: all set

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**